### PR TITLE
enh: remove churned members and workspaces from CIO

### PR DESCRIPTION
**Pull request by fontanierh**
## Description

https://github.com/dust-tt/tasks/issues/756

When we scrub the workspace (14 days after subscription is ended), we need to cleanup customerio.

What we're doing here is that we fetch all the workspace members, and for each of them we delete them from CIO if they do not have another workspace with an active subscription. This is imperfect, because if the other workspace of the user is churned but not yet scrubbed, we'll still delete the user when the first workspace is scrubbed. But this is relatively unimportant and also really an edge case.

We then also delete the workspace from CIO.

## Risk

Not too risky. Worst case we break the scrubbing but we'll find out.

### CODE REVIEW

File: front/temporal/scrub_workspace/activities.ts [hunk 1]
```diff
@@ -132,3 +141,113 @@ async function deleteDatasources(auth: Authenticator) {
...
+  }
+
+  // Fetch all the memberships for the workspace.
+  const workspaceMemberships = await MembershipResource.getLatestMemberships({
+    workspace: w,
+  });
+
+  // Fetch all the users in the workspace.
+  const userIds = workspaceMemberships.map((m) => m.userId);
+  const users = (
```
Comment (flvndvd) [inline front/temporal/scrub_workspace/activities.ts:159]: You could bundle this change in the MembershipResource conditioned by a `{includeUser: boolean}`.

File: front/temporal/scrub_workspace/activities.ts [hunk 2]
```diff
@@ -132,3 +141,113 @@ async function deleteDatasources(auth: Authenticator) {
...
+  const workspaceSids = Object.values(workspaceById).map((ws) => ws.sId);
+  const subscriptionsByWorkspaceSid = workspaceSids.length
+    ? await subscriptionForWorkspaces(workspaceSids)
+    : {};
+
+  // Process the workspace users in chunks of 16.
+  const chunks = chunk(users, 16);
+
+  for (const c of chunks) {
+    await Promise.all(
```
Comment (flvndvd) [inline front/temporal/scrub_workspace/activities.ts:204]: IMGO no need to have concurrency here.

**Review by flvndvd — COMMENTED** [review]
Definitely not a big fan of using the raw models, can we at least move most of this logic in `front/lib/xxx` so we can easily refactor once we have all the resources 🙏

File: front/temporal/scrub_workspace/activities.ts (hunk 1 above)
Comment (fontanierh) [reply front/temporal/scrub_workspace/activities.ts:159]: eeeeeeh yeah i could

Comment (fontanierh) [reply front/temporal/scrub_workspace/activities.ts:159]: but if this is the only place we'll ever use it, do we want it ?

File: front/temporal/scrub_workspace/activities.ts (hunk 2 above)
Comment (fontanierh) [reply front/temporal/scrub_workspace/activities.ts:204]: Why not ?

**Comment by fontanierh** [comment]
@flvndvd fair

> can we at least move most of this logic in front/lib/xxx

File: front/temporal/scrub_workspace/activities.ts [hunk 3]
```diff
@@ -1,17 +1,26 @@
-import { ConnectorsAPI } from "@dust-tt/types";
+import { ConnectorsAPI, removeNulls } from "@dust-tt/types";
 import { chunk } from "lodash";
+import _ from "lodash";
```
Comment (flvndvd) [inline front/temporal/scrub_workspace/activities.ts:3]: Can we harmonize imports, here?

File: front/temporal/scrub_workspace/activities.ts (hunk 2 above)
Comment (flvndvd) [reply front/temporal/scrub_workspace/activities.ts:191]: Bad experience lately with the `Promise.all` 🥵

**Review by flvndvd — APPROVED** [review]
LGTM.

File: front/temporal/scrub_workspace/activities.ts (hunk 3 above)
Comment (fontanierh) [reply front/temporal/scrub_workspace/activities.ts:3]: sure i'll remove `{ chunk }`

File: front/temporal/scrub_workspace/activities.ts (hunk 2 above)
Comment (fontanierh) [reply front/temporal/scrub_workspace/activities.ts:191]: I'll meet you halfway and reduce to 4 :p

**Pull request merged**